esp-hal/src/timer/systimer.rs

```diff
@@ -396,20 +396,20 @@ pub trait Comparator {
     fn set_enable(&self, enable: bool) {
         let systimer = unsafe { &*SYSTIMER::ptr() };
 
-        #[cfg(not(esp32s2))]
         critical_section::with(|_| {
+            #[cfg(not(esp32s2))]
             systimer.conf().modify(|_, w| match self.channel() {
                 0 => w.target0_work_en().bit(enable),
                 1 => w.target1_work_en().bit(enable),
                 2 => w.target2_work_en().bit(enable),
                 _ => unreachable!(),
             });
-        });
 
-        #[cfg(esp32s2)]
-        systimer
-            .target_conf(self.channel() as usize)
-            .modify(|_r, w| w.work_en().bit(enable));
+            #[cfg(esp32s2)]
+            systimer
```

Comment:
Why move this into the critical section? The register isn't shared on the S2.

Reply:
The better question is, why wasn't this (and, like all other operations) in a critical section? Every method needs a shared reference only, what prevents users from shooting themselves in the foot by doing things in interrupts that they shouldn't?

Reply:
Alarm is Sync as things currently stand :)

Reply:
Why would you prefer `!Sync` over `&mut self` methods?

Reply:
I think `&mut self` is artificially restrictive. At the end of the day the volatile registers are basically glorified `Cell`s.
I really don't want to have to pull out `RefCell`, `Cell`, etc. to share what is already `Cell` amongst objects in the same thread.
imo, if an exclusive reference isn't needed it shouldn't be required.
If users want to enforce an invariant that requires exclusive access then they can use the Rust tool for that, either by taking ownership of the object or taking a mutable reference to the object. (This is done in `FronzenUnit` for example)

Side note there's also [this part](https://github.com/esp-rs/esp-hal/blob/main/documentation/API-GUIDELINES.md#api-surface) of the API GUIDELINES.
> Avoiding `&mut self` when `&self` is safe to use. `&self` is generally easier to use as an API. Typical applications of this are where the methods just do writes to registers which don't have side effects.

(Though I suppose pointing it out doesn't matter since we're questioning the guideline itself)